File: src/node/magma/MutableDivideState.ts
/*import java.util.ArrayList;*/
/*import java.util.Collection;*/
/*import java.util.Optional;*/
/*import java.util.stream.Stream;*/
/**/class MutableDivideState implements DivideState {
	private readonly segments : /*Collection<String>*/ = /*new ArrayList<>*/();
	/*private final CharSequence input*/;
	private index : /*int*/ = /* 0*/;
	private buffer : /*StringBuilder*/ = /*new StringBuilder*/();
	private depth : /*int*/ = /* 0*/;
	/*MutableDivideState(final CharSequence input) {
        this.input = input;
    }*/
	advance : /*DivideState*/(){
		/*this.segments.add(this.buffer.toString());*/
		/*this.buffer = new StringBuilder();*/
		/*return this;*/
	}
	append : /*DivideState*/(/*final char c*/){
		/*this.buffer.append(c);*/
		/*return this;*/
	}
	stream : /*Stream<String>*/(){
		/*return this.segments.stream();*/
	}
	isLevel : /*boolean*/(){
		/*return 0 == this.depth;*/
	}
	enter : /*DivideState*/(){
		/*this.depth++;*/
		/*return this;*/
	}
	exit : /*DivideState*/(){
		/*this.depth--;*/
		/*return this;*/
	}
	pop : /*Character>>*/(){
		/*if (this.index >= this.input.length()) return Optional.empty();*/
		/*final var c = this.input.charAt(this.index);*/
		/*this.index++;*/
		/*return Optional.of(new Tuple<>(this, c));*/
	}
	popAndAppendToTuple : /*Character>>*/(){
		/*return this.pop().map(tuple -> {
            final var appended = tuple.left().append(tuple.right());
            return new Tuple<>(appended, tuple.right());
        }*/
		/*);*/
	}
	popAndAppendToOption : /*Optional<DivideState>*/(){
		/*return this.popAndAppendToTuple().map(Tuple::left);*/
	}
	isShallow : /*boolean*/(){
		/*return 1 == this.depth;*/
	}
	/**/}/**/

File: src/java/magma/rule/divide/divide/MutableDivideState.java
package magma.rule.divide.divide;

import java.util.ArrayList;
import java.util.Collection;
import java.util.stream.Stream;

public final class MutableDivideState implements DivideState {
	private final Collection<String> segments = new ArrayList<>();
	private final StringBuilder buffer = new StringBuilder();
	private int depth = 0;

	@Override
	public Stream<String> stream() {
		return this.segments.stream();
	}

	@Override
	public DivideState advance() {
		this.segments.add(this.buffer.toString());
		this.buffer.setLength(0);
		return this;
	}

	@Override
	public DivideState append(final char c) {
		this.buffer.append(c);
		return this;
	}

	@Override
	public boolean isLevel() {
		return 0 == this.depth;
	}

	@Override
	public DivideState enter() {
		this.depth++;
		return this;
	}

	@Override
	public DivideState exit() {
		this.depth--;
		return this;
	}
}